### PR TITLE
DOC: expand the `scipy.interpolate` tutorial

### PR DESCRIPTION
**Pull request by ev-br**
Restructure and expand the tutorial for `scipy.interpolate`:

- Group the functionality into, roughly, 1D -- N-D gridded -- N-D unstructured data. Make smoothing separate.
- For 1D routines stress newer routines, CubicSpline et al
- Tone down `interp1d` which used to be featured as *the* way; this PR goes as far as declare `interp1d` legacy in the docs (let's keep the discussion on formalizing legacy API separate, this docs can be adjusted easily)
- Briefly discuss what FITPACK does, stress that it does smoothing not interpolation. Give pointers to how `splrep` et al interoperates with BSpline.
- Try stressing that `griddata` is not for gridded data :-).

There are several other things that need to be written, but this one already gets too large. So sending it as is, to serve as a basis for working on doc issues in `scipy.interpolate` and `documentation` labels.

Would appreciate reviews on all levels: from the overall structure and content down to grammar and text flow.

### CODE REVIEW

**Comment by lobpcg** [comment]
It's generally best to have all tests passing before asking for a review. Your PR has errors as, e.g., indicated by  
scipy.scipy (Main refguide_asv_check) Failing after 42m — Main refguide_asv_check failed
specifically from https://dev.azure.com/scipy-org/df8ffb44-7275-43bc-90f9-aa5c93ef5c51/_apis/build/builds/19579/logs/148 

2022-07-27T10:52:20.5972324Z File "doc/source/tutorial/interpolate.rst", line 198, in /home/vsts/work/1/s/tools/../doc/source/tutorial/interpolate.rst
2022-07-27T10:52:20.5972990Z Failed example:
2022-07-27T10:52:20.5974591Z     dspl.root() / np.pi
2022-07-27T10:52:20.5975299Z Exception raised:
2022-07-27T10:52:20.5975836Z     Traceback (most recent call last):
2022-07-27T10:52:20.5976468Z       File "/opt/hostedtoolcache/Python/3.9.13/x64/lib/python3.9/doctest.py", line 1334, in __run
2022-07-27T10:52:20.5977140Z         exec(compile(example.source, filename, "single",
2022-07-27T10:52:20.5977832Z       File "<doctest /home/vsts/work/1/s/tools/../doc/source/tutorial/interpolate.rst[46]>", line 1, in <module>
2022-07-27T10:52:20.5978445Z         dspl.root() / np.pi
2022-07-27T10:52:20.5979313Z     AttributeError: 'CubicSpline' object has no attribute 'root'
2022-07-27T10:52:20.5979759Z 
2022-07-27T10:52:20.5980006Z 
2022-07-27T10:52:20.5980677Z File "doc/source/tutorial/interpolate.rst", line 209, in /home/vsts/work/1/s/tools/../doc/source/tutorial/interpolate.rst
2022-07-27T10:52:20.5981338Z Failed example:
2022-07-27T10:52:20.5981818Z     dspl.root(extrapolate=False) / np.pi
2022-07-27T10:52:20.5982312Z Exception raised:
2022-07-27T10:52:20.5982784Z     Traceback (most recent call last):
2022-07-27T10:52:20.5983419Z       File "/opt/hostedtoolcache/Python/3.9.13/x64/lib/python3.9/doctest.py", line 1334, in __run
2022-07-27T10:52:20.5984070Z         exec(compile(example.source, filename, "single",
2022-07-27T10:52:20.5984748Z       File "<doctest /home/vsts/work/1/s/tools/../doc/source/tutorial/interpolate.rst[47]>", line 1, in <module>
2022-07-27T10:52:20.5985413Z         dspl.root(extrapolate=False) / np.pi
2022-07-27T10:52:20.5986245Z     AttributeError: 'CubicSpline' object has no attribute 'root'
2022-07-27T10:52:20.5986667Z 
2022-07-27T10:52:20.5986931Z 
2022-07-27T10:52:20.5987529Z File "doc/source/tutorial/interpolate.rst", line 216, in /home/vsts/work/1/s/tools/../doc/source/tutorial/interpolate.rst
2022-07-27T10:52:20.5988187Z Failed example:
2022-07-27T10:52:20.5988663Z     dspl.solve(0.5, extrapolate=False)
2022-07-27T10:52:20.5989144Z Expected:
2022-07-27T10:52:20.5989637Z     array([0.33332755, 1.66667195, 2.3333271])
2022-07-27T10:52:20.5990106Z Got:
2022-07-27T10:52:20.5990587Z     array([1.04717938, 5.23600436, 7.33036327])
2022-07-27T10:52:20.5990923Z 
2022-07-27T10:52:20.5991156Z 
2022-07-27T10:52:20.5991762Z File "doc/source/tutorial/interpolate.rst", line 309, in /home/vsts/work/1/s/tools/../doc/source/tutorial/interpolate.rst
2022-07-27T10:52:20.5992388Z Failed example:
2022-07-27T10:52:20.5993053Z     der.k
2022-07-27T10:52:20.5993462Z Expected:
2022-07-27T10:52:20.5993905Z     3
2022-07-27T10:52:20.5994503Z Got:
2022-07-27T10:52:20.5994976Z     2
2022-07-27T10:52:20.5995654Z /home/vsts/work/1/s/tools/refguide_check.py:673: DeprecationWarning: elementwise comparison failed; this will raise an error in the future.
2022-07-27T10:52:20.5996352Z   if want == got:

**Comment by ev-br** [comment]
OK, doctests should be fixed now, the `_stats.pyx` failure is very much unrelated. So reviewing this should be safe now. If there are further doctest failures, I'll fix them up all right :-).


```
@cython.wraparound(False)
@cython.boundscheck(False)
def _weightedrankedtau(ordered[:] x, ordered[:] y, intp_t[:] rank, weigher, bool additive):
^
------------------------------------------------------------

/home/circleci/repo/scipy/stats/_stats.pyx:173:0: Referring to a memoryview typed argument directly in a nested closure function is not supported in Cython 0.x. Either upgrade to Cython 3, or assign the argument to a local variable and use that in the nested function.
```

File: doc/source/tutorial/interpolate.rst
Comment (mdhaber) [inline doc/source/tutorial/interpolate.rst:351]: I got a bit lost in here. Is it assumed that the reader knows what a "knot" is? 
There seems to be a lot of detail here about how the b-splines are working internally without a punchline (how the user can use it to do something).

File: doc/source/tutorial/interpolate.rst
Comment (mdhaber) [inline doc/source/tutorial/interpolate.rst:382]: "and is zero outside of these knots"
Maybe this point would be illustrated better if `a = 0` and `b = 5` (fixed)? Right now, I can't see what the values are outside the interval.

File: doc/source/tutorial/interpolate.rst
Comment (mdhaber) [inline doc/source/tutorial/interpolate.rst:403]: After reading this, I'm guessing that this is the main point about why to use b-splines:

> The b-spline basis generally more computationally stable then the power basis

Maybe show an example of that? What are the downsides (computation time?)?

File: doc/source/tutorial/interpolate.rst
Comment (mdhaber) [inline doc/source/tutorial/interpolate.rst:411]: It looks like there is no special functionality for making spline curves; it is just a way of using the existing functions? The different parameterizations are implemented by the user? Maybe mention it.

**Review by mdhaber — COMMENTED** [review]
Took a look at the 1D stuff; this is a nice addition. 
This is getting a little long. Consider breaking it up into separate pages - if not only for the sake of the reader, help the browser out. I think it sometimes forgets to render the math by the time it finishes the rest of the page.

File: doc/source/tutorial/interpolate.rst
Comment (ev-br) [reply doc/source/tutorial/interpolate.rst:37]: Indeed, I'm not sure where to put it. Specific suggestions? I guess it's tied to a restucture/split into separate pages (am *very* open to suggestions.)

File: doc/source/tutorial/interpolate.rst
Comment (ev-br) [reply doc/source/tutorial/interpolate.rst:57]: There are, too many of them in fact :-). `interp1d(x, y, kind="linear")` or `make_interp_spline(x, y, k=1)` [which is what `interp1d(x, y, kind="linear")`, a third equivalent way, delegates to under the hood.]  Do you think any of these are helpful this early in the page?

File: doc/source/tutorial/interpolate.rst
Comment (ev-br) [reply doc/source/tutorial/interpolate.rst:190]: Not sure what you mean, mind expanding? If you have a preference on how to structure / explain this, I'd go with it. 
Or maybe it's a sign of this whole subsection being extraneous?

File: doc/source/tutorial/interpolate.rst
Comment (ev-br) [reply doc/source/tutorial/interpolate.rst:214]: Thanks! Will add both suggestions in the next batch of tweaks.

File: doc/source/tutorial/interpolate.rst
Comment (ev-br) [reply doc/source/tutorial/interpolate.rst]: Thanks, changed.

File: doc/source/tutorial/interpolate.rst
Comment (ev-br) [reply doc/source/tutorial/interpolate.rst]: Meant to mean as a continuation from the previous paragraph: "Note that by default `make_interp_spline` constructs a cubic spline... this way, the default result of ... is equivalent to ... ", but that obviously did not work. Tried changing it, would appreciate a read-over or a copy-edit to read better.

File: doc/source/tutorial/interpolate.rst
Comment (ev-br) [reply doc/source/tutorial/interpolate.rst:353]: Sigh. Need to think how to restructure this whole section. Tried to avoid going into a lengthy teaching session, and wiggle my way through at least two sets of readers (those who see it for the first time and those who look for details), trying to document our spline format conventions (there are several in the literature and details need to be fished out, which is frustrating) and giving an overview of the functionality. Clearly, this was not a success :-).

File: doc/source/tutorial/interpolate.rst
Comment (ev-br) [reply doc/source/tutorial/interpolate.rst:384]: Sadly, these things extrapolate by default, so no, it might be even more confusing. I guess I'll need to say it in the text.

File: doc/source/tutorial/interpolate.rst
Comment (ev-br) [reply doc/source/tutorial/interpolate.rst:405]: Don't have a handly example off the bat, not sure there is one without going into *gory* details of what Nikolay called *the shady world of b-splines* when he wrote the CubicSpline code :-). Will think of ways of rephrasing. Maybe need to stress that the localized nature of the basis  often makes linear algebra banded.

File: doc/source/tutorial/interpolate.rst
Comment (ev-br) [reply doc/source/tutorial/interpolate.rst:413]: To make it more confusing, there is `splprep`  below in the smoothing splines section.

**Comment by ev-br** [comment]
Thanks Matt! Re splitting it --- and yes this needs splitting --- I'm not very up-to-date of our infrastructure for multiple rst pages. It used to be basically a single rst tutorial per submodule, do we have a better way now?

**Comment by mdhaber** [comment]
> It used to be basically a single rst tutorial per submodule, do we have a better way now?

Very good question. Currently stats is broken out; see `stats.rst` and the `rst` files in the `stats` folder [here](https://github.com/scipy/scipy/tree/main/doc/source/tutorial).

We're also working on allowing notebooks for tutorials like NumPy does. See https://github.com/scipy/scipy/pull/16699#issuecomment-1199940486.

File: doc/source/tutorial/interpolate.rst
Comment (mdhaber) [reply doc/source/tutorial/interpolate.rst:37]: Yeah let's wait on figuring out a better place. But if this were all going on one page, I'd just put it at the end. 

Alternatively, if you have any suggestions for how someone _can_ work with missing data, and if we split into multiple pages, maybe you could expound on that in its own page.

File: doc/source/tutorial/interpolate.rst
Comment (mdhaber) [reply doc/source/tutorial/interpolate.rst:57]: Hmm I thought it would be simpler than cubic, but it's not. Up to you - is it worth mentioning the `k` parameter of `make_interp_spline`? You talk about the degree of B-splines `k`, but it's not passed as an argument anywhere.

File: doc/source/tutorial/interpolate.rst
Comment (mdhaber) [reply doc/source/tutorial/interpolate.rst:190]: In the part above:
```
A `CubicSpline` object's ``__call__`` method accepts both scalar values and
arrays. It also accepts a second argument, ``nu``, to evaluate the 
derivative of order ``nu``. As an example, we plot the derivatives of a spline:
```

I was surprised that you could use the `__call__` method and pass a cryptically-named keyword argument to evaluate the derivative of the spline. It doesn't seem like as good API design as this object-oriented `derivative` method here, so I think we should emphasize the use of the `derivative` method instead. The suggestion is to eliminate the discussion of the `nu` keyward argument of `__call__`, and just talk about derivatives down here.

File: doc/source/tutorial/interpolate.rst
Comment (mdhaber) [reply doc/source/tutorial/interpolate.rst]: Something I'm understanding now that I didn't understand before is that `CubicSpline` returns a `PPoly`, and `make_interp_spline` returns a `BSpline`.

Maybe you can help me fill in other blanks and I can suggest what might have helped me understand the first time.

It seems that `PPoly` and `BSpline` are the two fundamental spline representations that you want to cover in the tutorial, right?

Is there a function such that <function> is to `PPoly` as `make_interp_spline` is to `BSpline`? `CubicSpline` doesn't fulfill this relationship because it can't be of arbitrary order.

File: doc/source/tutorial/interpolate.rst
Comment (mdhaber) [reply doc/source/tutorial/interpolate.rst:353]: > Tried to avoid going into a lengthy teaching session, and wiggle my way through at least two sets of readers 

Yeah, I'm the kind who sees it for the first time.

I'm not sure how to do both at once. I feel like the API reference is for those who know the theory, and tutorial is for those who don't.

File: doc/source/tutorial/interpolate.rst
Comment (j-bowhay) [inline doc/source/tutorial/interpolate.rst:365]:
```suggestion
The b-spline basis are generally more computationally stable then the power basis
```

**Review by j-bowhay — COMMENTED** [review]
Some very minor grammar and formatting points, feel free to ignore :smiley:

File: doc/source/tutorial/interpolate.rst
Comment (ev-br) [reply doc/source/tutorial/interpolate.rst:1052]: IIRC, having a non-whitespace symbol right after the backtick is (was?) confusing sphinx. Would be nice to double-check

**Comment by ev-br** [comment]
OK, am coming back to this one. Reworked the 1D section, broken up the document into separate pieces, as requested. Does it render better now @mdhaber ? Would appreciate a review!

File: doc/source/tutorial/interpolate.rst
Comment (ev-br) [reply doc/source/tutorial/interpolate.rst]: Moved this to the bottom of the '1D interpolation' section.

File: doc/source/tutorial/interpolate.rst
Comment (ev-br) [reply doc/source/tutorial/interpolate.rst]: @vnmabus no, it is not similar. `np.interp` is generally fastest, and is in fact recommended. `interp1d` has the most overhead and is not recommended. `make_interp_spline(...., k=1)`  should be comparable to `np.interp` if somewhat slower. This is probably useful is limited contexts: when you want flexibility in `k=...` or if you insist on passing around a linearly interpolating callable object.

File: doc/source/tutorial/interpolate.rst
Comment (ev-br) [reply doc/source/tutorial/interpolate.rst]: Well, I respectfully disagree. Without going into a discussion of what is and what is not a good design, I'd like to somehow demonstrate both ways in the tutorial. The main difference is that `spl(x, nu=2)` evaluates the second derivative in-place, and `spl.derivative(2)` constructs a new object of a degree two smaller which represents the second derivative. I'm open to suggestions of how to tone down the use of the `nu` argument if you feel strongly about it, but some demonstration I'd like to keep.

File: doc/source/tutorial/interpolate.rst
Comment (ev-br) [reply doc/source/tutorial/interpolate.rst]: Is it better now that the page is broken up?

File: doc/source/tutorial/interpolate.rst
Comment (ev-br) [reply doc/source/tutorial/interpolate.rst]: The "this way" thing now reads "By default, the result of `make_interp_spline...", does it read better?

The difference between PPoly and BSpline : I tried to expand the section now that it's separate.

File: doc/source/tutorial/interpolate.rst
Comment (ev-br) [reply doc/source/tutorial/interpolate.rst]: > Yeah, I'm the kind who sees it for the first time.

Then you are a perfect reader of this! Seriously.

Mixing up `np.Polynomial` here would be a bit confusing I suspect. Note that docstrings of both `PPoly` and `BSpline` list pure-python illustrations. I tried to refer to the docstrings, let me know if it worked. Or would it help to copy these implementations to the tutorial?

File: doc/source/tutorial/interpolate.rst
Comment (ev-br) [reply doc/source/tutorial/interpolate.rst]: I don't think there is a general guidance which one to use (as in, a theorem). Fritsch and Carlson paper, referenced in the pchip docstring, gives a whole family of these things, with the motivation of producing curves which are "smooth and pleasing to an eye" (quoting from memory, but quoting nonetheless). So, up to you. I'm typically using pchip, but that's an arbitrary decision.

As to why aren't they a `MonotoneInterpolant` with methods... Historic reasons. And we won't be adding more wrappers, there are too many of those already. (there is also an argument attributed to GvR that if you have a `method=...` parameter with two values, you probably need two functions).

File: doc/source/tutorial/interpolate.rst
Comment (ev-br) [reply doc/source/tutorial/interpolate.rst]: Thanks! Moved to separate page.

File: doc/source/tutorial/interpolate.rst
Comment (ev-br) [reply doc/source/tutorial/interpolate.rst]: >  I had in my mind "function = old low level API" so I didn't even pay attention to this one

Well, this is very definitely *not* how scipy.interpolate operates! 
While this misconception exists, I would very much appreciate suggestions of how to communicate that it is, in fact, a misconception.

As to why --- Not everything needs to be a class. Were we designing it from scratch, I'd probably advocate for a handful of bases classes and factory functions returning instances of these classes. But this is moot, and the argument of "no more wrappers" applies.

File: doc/source/tutorial/interpolate.rst
Comment (mdhaber) [reply doc/source/tutorial/interpolate.rst]: Oops, my word choice needed improvement there; sorry about that. I didn't notice the distinction before, so perhaps it's it can be emphasized more if the difference in performance is substantial.

File: doc/source/tutorial/interpolate/ND_unstructured.rst
Comment (j-bowhay) [inline doc/source/tutorial/interpolate/ND_unstructured.rst:72]:
```suggestion
    `griddata` is based on triangulation, hence is appropriate for unstructured,
```

**Review by j-bowhay — COMMENTED** [review]
A few suggestions after proof reading

File: doc/source/tutorial/interpolate/ND_regular_grid.rst
Comment (ev-br) [reply doc/source/tutorial/interpolate/ND_regular_grid.rst:68]: Not sure about that actually. The intent was that if you (a user) prefer to use a convenience function instead of explicitly creating a class instance and then calling it, then does just that `interpn`. Would appreciate a suggestion on how to reword this to avoid confusion.

**Comment by ev-br** [comment]
Thanks for the proofreading Jake!

File: doc/source/tutorial/interpolate/ND_regular_grid.rst
Comment (j-bowhay) [reply doc/source/tutorial/interpolate/ND_regular_grid.rst:68]: Yes you're right, misread the intent here as I reading through it quickly. How about:
```suggestion
If you prefer a functional interface opposed to explicitly creating a class instance,
```

Comment (ev-br) [reply doc/source/tutorial/interpolate/ND_regular_grid.rst]: Thanks Jake! Your rephrase is in :-).

**Comment by ev-br** [comment]
All right, this PR is ready from my side. The last commits expand on the 2D FITPACK spline wrappers, stressing the difference between scattered and gridded problems (this part I was sorely missing, FWIW). The only thing left is to port the RBF part from the legacy Rbf class usage to the new RBFInterpolator, but this can be done in a follow-up.

File: doc/source/tutorial/interpolate.rst
Comment (mdhaber) [inline doc/source/tutorial/interpolate.rst:23]:
```suggestion
+------------------+-------------------------+------------------------------+------------------------+---------------------------------------+
|                  | **kind**                | **routine**                  |  **continuity**        | **comment**                           |
+==================+=========================+==============================+========================+=======================================+
```

File: doc/source/tutorial/interpolate/1D.rst
Comment (mdhaber) [inline doc/source/tutorial/interpolate/1D.rst:115]: "As an illustration" of what, though?
There are nice transition sentences at the top of the first few sections. It would be nice to have just one sentence like that to introduce why B-splines are needed.

**Review by mdhaber — APPROVED** [review]
Thanks @ev-br. I really like the table, which was not there in my first pass, and the new organization across several files makes it a lot more readable. 

Just two minor suggestions; I figured I'd wait to commit in case you have any last things. But after that, I'd be happy to merge if you think it's time.

File: doc/source/tutorial/interpolate/1D.rst
Comment (ev-br) [inline doc/source/tutorial/interpolate/1D.rst:114]:
```suggestion

B-splines form an alternative (if formally equivalent) representation of piecewise polynomials. This basis is generally more computationally stable than the power basis and is useful for a variety of applications which include interpolation, regression and curve representation. Details are given in the :ref:`piecewise polynomials section <tutorial-interpolate_ppoly>`, and here we illustrate their usage by constructing the interpolation of a sine function: 

```

File: doc/source/tutorial/interpolate/1D.rst
Comment (ev-br) [inline doc/source/tutorial/interpolate/1D.rst:115]:
```suggestion
```

File: doc/source/tutorial/interpolate/1D.rst
Comment (ev-br) [inline doc/source/tutorial/interpolate/1D.rst:116]:
```suggestion
```

File: doc/source/tutorial/interpolate/1D.rst
Comment (ev-br) [inline doc/source/tutorial/interpolate/1D.rst:117]:
```suggestion
```

File: doc/source/tutorial/interpolate/1D.rst
Comment (ev-br) [reply doc/source/tutorial/interpolate/1D.rst]: Thanks Matt! Tried to reword the paragraph in the last commit, does it read better?

**Comment by ev-br** [comment]
> Just two minor suggestions; I figured I'd wait to commit in case you have any last things. But after that, I'd be happy to merge if you think it's time.

Thank you! I merged your suggestions, which are a clear improvement IMO,  together with a rewording of the lead-up to b-splines. I think this PR is too large already, so I'll keep my other tweaks and additions to a follow-up or several (the first one is gh-17299, would asking to review that be too much?).

**Comment by mdhaber** [comment]
@ev-br I'll merge this, but I wanted to check with you - in stats these days, we typically squash merge (noting co-authors in the commit message) unless the separate commits are clearly well-defined. I would be tempted to do the same here (no coauthors, really), but I wanted to check with you before I do that. Feel free to rebase if you'd like and I can do a merge commit.

**Comment by ev-br** [comment]
Agreed, there's nothing worth keeping my in the commit his story here

**Comment by mdhaber** [comment]
Thanks @ev-br @j-bowhay!

**Pull request merged**
